In [ ]:
# Aquí comenzaremos a trabajar código en Python.

In [ ]:
# Implementación de las librerias utilizadas.
from IPython.display import display # Para mostrar múltiples resultados en una sola celda.
import pandas as pd
import numpy as np
import gdown

In [ ]:
#Se hace el montaje del Drive para el acceso a la DB
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#Aquí generé el acceso a la base de datos (CSV)
ruta_csv = '/content/drive/My Drive/Bedu_Tecnolochicas/cybersecurity_attacks.csv'

Hacemos un poco de "limpieza a la DB".
- Se reemplazan los datos NaN por cero.
- Se eliminan espacios en los encabezados de las columnas por guión bajo.
- Se eliminan espacios en blanco que no sean necesarios.
- Estandarizamos todo en minusculas para evitar errores.


In [ ]:
#Declaración de las variables generales
df = pd.read_csv(ruta_csv)

# Reemplazamos los NaN con un valor, en este caso se optó por: 0
df.fillna(0, inplace=True)

# Cambiar los encabezados de las columnas
df.columns = df.columns.str.replace(' ', '_')

# Eliminar espacios en blanco en todas las columnas de texto
df = df.applymap(lambda x: x.strip() if isinstance(x, str) else x)

# Convertir todo a minúsculas en todas las columnas de texto
df = df.applymap(lambda x: x.lower() if isinstance(x, str) else x)

<ipython-input-7-19a9f56383b0>:11: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df = df.applymap(lambda x: x.strip() if isinstance(x, str) else x)
<ipython-input-7-19a9f56383b0>:14: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df = df.applymap(lambda x: x.lower() if isinstance(x, str) else x)


In [ ]:
#Desplegamos la DB
display(df)

Timestamp Source_IP_Address Destination_IP_Address  \
0      2023-05-30 06:33:58     103.216.15.12           84.9.164.252   
1      2020-08-26 07:08:30    78.199.217.198         66.191.137.154   
2      2022-11-13 08:23:25      63.79.210.48          198.219.82.17   
3      2023-07-02 10:38:46     163.42.196.10        101.228.192.255   
4      2023-07-16 13:11:07     71.166.185.76        189.243.174.238   
...                    ...               ...                    ...   
39995  2023-05-26 14:08:42      26.36.109.26         121.100.75.240   
39996  2023-03-27 00:38:27      17.21.163.81         196.108.134.78   
39997  2022-03-31 01:45:49     162.35.217.57            98.107.0.15   
39998  2023-09-22 18:32:38    208.72.233.205         173.79.112.252   
39999  2023-10-10 11:59:52     14.102.21.108           109.198.45.7   

       Source_Port  Destination_Port Protocol  Packet_Length Packet_Type  \
0            31225             17616     icmp            503        data   
1            17245             48166     icmp           1174        data   
2            16811             53600      udp            306     control   
3            20018             32534      udp            385        data   
4             6131             26646      tcp           1462        data   
...            ...               ...      ...            ...         ...   
39995        31005              6764      udp           1428     control   
39996         2553             28091      udp           1184     control   
39997        22505             25152      udp           1043        data   
39998        20013              2703      udp            483        data   
39999        50137             55575     icmp           1175     control   

      Traffic_Type                                       Payload_Data  ...  \
0             http  qui natus odio asperiores nam. optio nobis ius...  ...   
1             http  aperiam quos modi officiis veritatis rem. omni...  ...   
2             http  perferendis sapiente vitae soluta. hic delectu...  ...   
3             http  totam maxime beatae expedita explicabo porro l...  ...   
4              dns  odit nesciunt dolorem nisi iste iusto. animi v...  ...   
...            ...                                                ...  ...   
39995         http  quibusdam ullam consequatur consequuntur accus...  ...   
39996         http  quaerat neque esse. animi expedita natus commo...  ...   
39997          dns  enim at aspernatur illum. saepe numquam eligen...  ...   
39998          ftp  officiis dolorem sed harum provident earum dis...  ...   
39999         http  eligendi omnis voluptate nihil voluptatibus do...  ...   

      Action_Taken  Severity_Level User_Information  \
0           logged             low    reyansh dugal   
1          blocked             low       sumer rana   
2          ignored             low     himmat karpe   
3          blocked          medium       fateh kibe   
4          blocked             low     dhanush chad   
...            ...             ...              ...   
39995       logged          medium      adira madan   
39996       logged            high        rati dara   
39997      blocked             low     samiha joshi   
39998      ignored             low    rasha chauhan   
39999       logged          medium      zaina kumar   

                                      Device_Information Network_Segment  \
0      mozilla/5.0 (compatible; msie 8.0; windows nt ...       segment a   
1      mozilla/5.0 (compatible; msie 8.0; windows nt ...       segment b   
2      mozilla/5.0 (compatible; msie 9.0; windows nt ...       segment c   
3      mozilla/5.0 (macintosh; ppc mac os x 10_11_5; ...       segment b   
4      mozilla/5.0 (compatible; msie 5.0; windows nt ...       segment c   
...                                                  ...             ...   
39995  mozilla/5.0 (ipad; cpu ipad os 14_2_1 like mac...       segment a   
39996  mozilla/5.0 (windows; u; windows 98; w

In [ ]:
#Se genera un panorama general de los datos. ¿Qué tanto tenemos para trabajar?
def info_general_df(df):
  print('Información general del DataFrame')
  print(f'Cantidad de filas y columnas: {df.shape}')
  print("-"*50)
  print(f'Cantidad de datos nulos por columna: \n{df.isnull().sum()}')
  print("-"*50)
  print(f'Cantidad de datos únicos por columna: \n{df.nunique()}')
  print("-"*50)
  print(f'Tipos de datos por columna: \n{df.dtypes}')

info_general_df(df)

Información general del DataFrame
Cantidad de filas y columnas: (40000, 25)
--------------------------------------------------
Cantidad de datos nulos por columna: 
Timestamp                 0
Source_IP_Address         0
Destination_IP_Address    0
Source_Port               0
Destination_Port          0
Protocol                  0
Packet_Length             0
Packet_Type               0
Traffic_Type              0
Payload_Data              0
Malware_Indicators        0
Anomaly_Scores            0
Alerts/Warnings           0
Attack_Type               0
Attack_Signature          0
Action_Taken              0
Severity_Level            0
User_Information          0
Device_Information        0
Network_Segment           0
Geo-location_Data         0
Proxy_Information         0
Firewall_Logs             0
IDS/IPS_Alerts            0
Log_Source                0
dtype: int64
--------------------------------------------------
Cantidad de datos únicos por columna: 
Timestamp                 39997


In [ ]:
print("Valores nulos: ")
print(df.isnull().sum())

Valores nulos: 
Timestamp                 0
Source_IP_Address         0
Destination_IP_Address    0
Source_Port               0
Destination_Port          0
Protocol                  0
Packet_Length             0
Packet_Type               0
Traffic_Type              0
Payload_Data              0
Malware_Indicators        0
Anomaly_Scores            0
Alerts/Warnings           0
Attack_Type               0
Attack_Signature          0
Action_Taken              0
Severity_Level            0
User_Information          0
Device_Information        0
Network_Segment           0
Geo-location_Data         0
Proxy_Information         0
Firewall_Logs             0
IDS/IPS_Alerts            0
Log_Source                0
dtype: int64


# **Se abordaron 5 preguntas las cuales se ponen de encabezado y se realiza el código enseguida para ir resolviendo la incognita **

**Análisis temporal de ciberataques:**
¿Hay patrones de tiempo específicos (diurnos, nocturnos, días de la semana) en los ataques según las marcas de tiempo?


In [ ]:
# Convertir la columna de Timestamp a datetime
# Se cambió 'ts' por 'Timestamp' que es el nombre correcto de la columna en la DB
df['timestamp'] = pd.to_datetime(df['Timestamp'], format='%Y-%m-%d %H:%M:%S.%f', errors='coerce')

# Extraer información temporal relevante
df['hour'] = df['timestamp'].dt.hour  # Extraer la hora
df['day_of_week'] = df['timestamp'].dt.dayofweek  # Extraer el día de la semana (0 = lunes, 6 = domingo)

# Clasificar ataques como diurnos o nocturnos
def clasificar_horario(hora):
    if 6 <= hora < 18:
        return 'diurno'
    else:
        return 'nocturno'

df['time_of_day'] = df['hour'].apply(clasificar_horario)

# Agrupar ataques por franja horaria (diurno/nocturno)
ataques_por_franja = df['time_of_day'].value_counts()

# Agrupar ataques por día de la semana
ataques_por_dia = df['day_of_week'].value_counts().sort_index()

# Mostrar resultados
print("Distribución de ataques por franja horaria:")
print(ataques_por_franja)

print("\nDistribución de ataques por día de la semana (0=Lunes, 6=Domingo):")
print(ataques_por_dia)

Distribución de ataques por franja horaria:
time_of_day
nocturno    40000
Name: count, dtype: int64

Distribución de ataques por día de la semana (0=Lunes, 6=Domingo):
Series([], Name: count, dtype: int64)


El análisis muestra que la mayoría de los ciberataques ocurren durante la noche con 40mil ataques. La distribución por días es constante, con ligeros picos los domingos, resaltando la necesidad de vigilancia continua y estrategias de seguridad en horarios críticos.

**Evaluación de protocolos inseguros:**
¿Qué protocolos están más frecuentemente asociados con ataques clasificados como de alto nivel de gravedad?


In [ ]:
# Convertimos los valores inválidos a NaT en lugar de generar errores.
# Verificamos que ya no existan valores NaT para Timestamp
df['Timestamp'] = pd.to_datetime(df['Timestamp'], errors='coerce', format='%Y-%m-%d %H:%M:%S')
print("Valores nulos en 'Timestamp':", df['Timestamp'].isnull().sum())

Valores nulos en 'Timestamp': 0


In [ ]:
#Eliminamos las filas con fechas inválidas
#Rellenamos con valores por defecto (No se me ocurrió de otra más que sea la fecha actúal)
df = df.dropna(subset=['Timestamp'])
from datetime import datetime
df['Timestamp'].fillna(datetime.now(), inplace=True)

<ipython-input-13-aefcc528c4d2>:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Timestamp'].fillna(datetime.now(), inplace=True)


In [ ]:
#Comprobamos que todo esté correcto (Nuevamente)
print(df['Timestamp'].head())  # Verifica que la columna esté limpia
print("Valores nulos restantes:", df['Timestamp'].isnull().sum())

0   2023-05-30 06:33:58
1   2020-08-26 07:08:30
2   2022-11-13 08:23:25
3   2023-07-02 10:38:46
4   2023-07-16 13:11:07
Name: Timestamp, dtype: datetime64[ns]
Valores nulos restantes: 0


In [ ]:
#Ahora que todo está "limpio" extraemos atributos
df['hour'] = df['Timestamp'].dt.hour
df['day_of_week'] = df['Timestamp'].dt.day_name()
df['time_of_day'] = df['hour'].apply(lambda x: 'nocturno' if x < 6 else 'diurno')

In [ ]:
# Primero filtraremos solo los ataques con alto nivel de severidad.
# Supongamos que los niveles de severidad son 'High', 'Medium', 'Low'
high_severity_attacks = df[df['Severity_Level'] == 'high'] # Changed 'data' to 'df' and 'Severity Level' to 'Severity_Level' to match the column name after cleaning

# Mostrar las primeras filas de los ataques de alta severidad
print(high_severity_attacks.head())

# Posterior a eso contaremos los protocolos más comunes en ataques de alta severidad.
protocol_counts = high_severity_attacks['Protocol'].value_counts().reset_index()
protocol_counts.columns = ['Protocol', 'Count']

# Por ultimo mostraremos los protocolos más frecuentes asociados a los ataques de mayor severidad.
print(protocol_counts.head(10))

             Timestamp Source_IP_Address Destination_IP_Address  Source_Port  \
6  2022-05-16 17:55:43     97.253.103.59           77.16.101.53        26562   
7  2023-02-12 07:13:17      11.48.99.245         178.157.14.116        34489   
8  2023-06-27 11:02:56     49.32.208.167           72.202.237.9        56296   
14 2023-02-01 13:17:17    54.163.130.178         62.112.149.214        47795   
16 2023-04-18 16:46:46     4.255.187.165        136.159.186.239        39934   

    Destination_Port Protocol  Packet_Length Packet_Type Traffic_Type  \
6              17416      tcp            379        data          dns   
7              20396     icmp           1022        data          dns   
8              20857      tcp           1281     control          ftp   
14             35243      udp            913        data          dns   
16              5259      tcp            838     control         http   

                                         Payload_Data  ... Network_Segment  \
6 

El análisis de los datos revela que los protocolos TCP, ICMP y UDP son los que con mayor frecuencia presentan ataques clasificados como de alta gravedad.
En particular, el protocolo TCP, al ser uno de los más utilizados para la transmisión de datos en redes de internet, se presenta como un objetivo recurrente en diversos tipos de ataques

**Identificación de puertos vulnerables:**
¿Cuáles son los puertos de destino más comúnmente atacados y qué tipo de tráfico se dirige a ellos?


In [ ]:
# Agrupar por Destination_Port y Traffic_Type para contar las ocurrencias
puertos_trafico = df.groupby(['Destination_Port', 'Traffic_Type']).size().reset_index(name='count')

# Ordenar los resultados por la cantidad de ocurrencias en orden descendente
puertos_trafico = puertos_trafico.sort_values(by='count', ascending=False)

# Mostrar los resultados
print(puertos_trafico)

# Agrupar solo por Destination_Port y contar las ocurrencias
puertos_mas_atacados = df['Destination_Port'].value_counts().reset_index()
puertos_mas_atacados.columns = ['Destination_Port', 'count']

# Mostrar los 10 puertos más atacados
print("Puertos más comúnmente atacados:")
print(puertos_mas_atacados.head(10))

# Filtrar los 10 puertos más atacados
top_puertos = puertos_mas_atacados.head(10)['Destination_Port']

# Filtrar el DataFrame original para los puertos más atacados
df_top_puertos = df[df['Destination_Port'].isin(top_puertos)]

# Agrupar por Destination_Port y Traffic_Type
trafico_por_puerto = df_top_puertos.groupby(['Destination_Port', 'Traffic_Type']).size().reset_index(name='count')

# Ordenar para visualización
trafico_por_puerto = trafico_por_puerto.sort_values(by=['Destination_Port', 'count'], ascending=[True, False])

# Mostrar los resultados
print("Tipo de tráfico dirigido a los puertos más atacados:")
print(trafico_por_puerto)



       Destination_Port Traffic_Type  count
16884             30804         http      4
31338             56672         http      4
33607             60616          ftp      4
34202             61717         http      4
24388             44330          dns      4
...                 ...          ...    ...
12691             23283         http      1
12690             23282         http      1
12689             23282          dns      1
12688             23281          dns      1
18118             32973          dns      1

[36237 rows x 3 columns]
Puertos más comúnmente atacados:
   Destination_Port  count
0             34117      6
1              7508      6
2             30804      5
3             37248      5
4             57508      5
5             56672      5
6             38787      5
7             36756      5
8             25889      5
9             39887      5
Tipo de tráfico dirigido a los puertos más atacados:
    Destination_Port Traffic_Type  count
0               7508  

Los puertos **34117** y **7508** son los más atacados.
El tráfico HTTP domina, seguido por DNS y FTP, lo que sugiere que los atacantes buscan vulnerabilidades en servicios web, resolución de nombres y transferencias de archivos.
La diversidad de tráfico en los puertos más atacados indica que los atacantes intentan múltiples enfoques para comprometer los servicios.

**Análisis de patrones geográficos:**
¿Qué regiones geográficas generan la mayor cantidad de ataques según las direcciones IP de origen y los datos de geolocalización?

In [ ]:
# Agruparemos por región y contaremos el número de ataques en cada uno.
region_analysis = df.groupby('Geo-location_Data').size().reset_index(name='Attack_Count')

# Ordenar las regiones de mayor a menor cantidad de ataques.
region_analysis = region_analysis.sort_values(by='Attack_Count', ascending=False)

# Dando como resultado: Mostrar las 10 regiones con más ataques
print(region_analysis.head(10))

               Geo-location_Data  Attack_Count
2784        ghaziabad, meghalaya            16
3959  kalyan-dombivli, jharkhand            15
2795      ghaziabad, uttarakhand            14
2793          ghaziabad, tripura            14
5402            motihari, odisha            13
681         aurangabad, nagaland            13
7725     srikakulam, uttarakhand            13
2778        ghaziabad, jharkhand            13
4572          kottayam, nagaland            13
519               amroha, sikkim            13


El análisis anterior revela que las regiones con la mayor cantidad de ataques
registrados son Ghaziabad, Meghalaya y Kalyan-Dombivli ,Jharkhand. Este patrón indica que las regiones de dónde provienen la mayor cantidad de ataques  están localizados en India, un país con una infraestructura tecnológica en expansión y un gran tráfico de datos

**Relación entre firmas de ataque y acciones tomadas:**
¿Qué firmas de ataque tienen una mayor probabilidad de desencadenar acciones específicas (como bloqueos, alertas, etc.)?



In [ ]:
# Agrupamos Attack_Signature y Action_Taken para contar cuántas veces ocurre cada combinación.

# Contar las ocurrencias de cada combinación
resumen = df.groupby(['Attack_Signature', 'Action_Taken']).size().reset_index(name='count')

# Calcular la proporción relativa por cada firma de ataque
resumen['proporcion'] = resumen['count'] / resumen.groupby('Attack_Signature')['count'].transform('sum')

# Convertir la proporción en porcentaje
resumen['proporcion_%'] = resumen['proporcion'] * 100

print(resumen)

  Attack_Signature Action_Taken  count  proporcion  proporcion_%
0  known pattern a      blocked   6806    0.339012     33.901176
1  known pattern a      ignored   6685    0.332985     33.298466
2  known pattern a       logged   6585    0.328004     32.800359
3  known pattern b      blocked   6723    0.337432     33.743224
4  known pattern b      ignored   6591    0.330807     33.080707
5  known pattern b       logged   6610    0.331761     33.176069


Para ambas firmas de ataque (known pattern a y known pattern b), las acciones tomadas (blocked, ignored, y logged) tienen proporciones muy similares.
Esto sugiere que no hay un sesgo evidente hacia una acción en particular para estas firmas de ataque. En otras palabras, el sistema reacciona de manera equilibrada ante estas firmas.

